### DISCUSSION TITLE
Direct Port Manipulation

### DISCUSSION THREAD

**Discussion by Randall-Scharpf** [SpenceKonde/megaTinyCore#347] [Future Development Discussion]
I've referred to the Port Manipulation reference on the Arduino documentation, and am intending to use the said methods in my latest project. Here is the page in question (https://www.arduino.cc/en/Reference/PortManipulation) and here is the code I've written:
```
DDRB  = (DDRB  & 0b11000000);
PORTB = (PORTB & 0b11000000) | (1 << column);
DDRB  = (DDRB  & 0b11000000) | (1 << column);
buttons = (buttons << 5) | ((PINB & SL[column]) >> 1) | (SR[column] & PINB);
```
I cannot compile my code when the Arduino IDE is set to `Boards > megaTinyCore > ATtiny3216/.../406 (Optiboot)`. I am intending to deploy to the ATtiny3216 using a Serial interface with another Arduino. I tested compiling the code with the board type set to Uno, and it worked, however I don't really care because I'm not using the Uno for my project, I simply used that fact to narrow down cause of the error (which is `'PORTB' was not declared in this scope` and a handful other similar errors) to a misunderstanding of this library. I've perused the datasheet and tried some other register names, but I am yet to find one that works. I would like to ask:
 - Is it possible to use port manipulation on the ATtiny3216 in the same way as I would on the Arduino Uno?
 - If it is possible, do I need new names for the variables, a new "#include" to make it work (I have none at the moment because I have found none that fix my issue), or a new method altogether?
 - Is there a page in the datasheet that makes the existence of this method / how to use it clear that I am missing, so I don't find myself similarly stumped in the future?
 
 Thanks.

**Comment by SpenceKonde** [comment]
The ATtiny3216 is a modern AVR, (specifically a tinyAVR 1-series). Like all modern AVR parts (all AVR devices released since 2016) they use a different structure for the port registers. On the"modern" AVR devices (not official word. I asked my contact in microchip what they call these two broad families of parts (pre-2016/classic and post=2016 woth all the new goodies). He indicated that they didn't have a word, but were talking about it internally because it was b ecoming evident that they need one, which I could have told them in 2014 when the new parts were still diagrams on whiteboards.)... those devices have 2 sets of registers. There are the "VPORTx registers - VPORTx.IN, VPORTx.OUT, VPORTx.DIR, VPORTx.INTFLAGS, and there are the full service PORTx registers

Writing to the VPORTx.IN toggles the VPORTx.OUT just like on classic devices (though the pullup is not controlled in the same way -for that you need to use the full fledged PORTx registers). 

The VPORTx registers are located in the "low I/O space" so you can do  VPORTA.OUT |= 1 << x; (where x is constant known at compiletime) and it will compile to a single sbi instruction which executes in a single clock cycle (if that same line is compiled for an unknown value of x that is between 0 and 7, it's like 20-24 bytes of flash and (Because of a mini loop in the middle) I counted around 40 clock cycles to execute it! 
PORTx are not - thought they have OUTSET, OUTCLR so that you can do atomic sets and clears (and toggles) - the advantage of using PORTx,OUTSET vs VPORTx.OUT |= is that you can do more than 1 bit at a time, and what you're setting doesn't need to be known at compile time (on the other hand, if it is, it's gonna be less efficient
The same thing with sbi also works on classic AVRs (though sbi and it's inverse, cbi, were 2-cycle instructions back on classic; now they are single-clock.. There's also a skipif instructiion that tests the bits in the low IO registers, so testing them is super-fast and efficient (it's a slap in the face that we have to waste half out SBIS/SBIC-able registers on VPORT.OUT and VPORT.DIR which we set ourselves, and 

Note that the ATtiny3216 is not related to ATTinyCore - it is supported by megaTinyCore over here - https://github.com/SpenceKonde/megaTinyCore/discussions and there is a page in the documentation going over this: https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/DirectPortManipulation.md
